### ISSUE TITLE
1.5.0 breaks webpack builds

### ISSUE THREAD

**Issue by gilbox** [foretagsplatsen/numbro#74]
I think this has to do with issue #62, although in that issue it says that the relevant code was merged in 1.4.0 even though 1.4.0 doesn't break webpack so I'm not really sure. 

Anyway, my theory is that because of some dynamic require webpack tries loading in all the files in the node_modules/numbro because that's how webpack handles dynamic requires.

The error is very long but it starts out like this:

```
WARNING in ./~/numbro/numbro.js
Critical dependencies:
970:60-112 the request of a dependency is an expression
 @ ./~/numbro/numbro.js 970:60-112
```

**Comment by BenjaminVanRyseghem** [comment]
Note that I often fuck up the deployment, so you should use the latest version, as it could be that the code was not actually in the version announced (npm does not let you overwrite a broken version :confused: )

**Comment by gilbox** [comment]
@BenjaminVanRyseghem, I don't think I made myself clear. `1.5.0` is broken but `1.4.0` works.

**Comment by BenjaminVanRyseghem** [comment]
oh :octopus: 

I will investigate then (but I am on vacations now so :smile:)

**Comment by gilbox** [comment]
I'm not too worried about it, just thought you'd like to know. Here is a partial solution:

```
    // CommonJS module is defined
    if (hasModule) {
        module.exports = numbro;
        require.ensure([], function(require) {
          // Load all languages
          var fs = require('fs'),
              path = require('path');
          var langFiles = fs.readdirSync(path.join(__dirname, 'languages'));
          langFiles.forEach(function (langFile) {
              numbro.language(path.basename(langFile, '.js'), require('./languages/' + langFile));
          });
        })
    }
```

However, it still causes the error:

```
Module not found: Error: Cannot resolve module 'fs'
```

This last error could perhaps be fixed using webpack `IgnorePlugin`, but it's a shame that you'd have to add this additional configuration just to use `numbro`. I'm not sure how else to work around it though.

**Comment by BenjaminVanRyseghem** [comment]
Maybe we could use something like [this](http://stackoverflow.com/questions/4224606/how-to-check-whether-a-script-is-running-under-node-js/24279593#24279593) to determine if the code is run under *nodejs*. What do you think?

**Comment by gilbox** [comment]
I don't think it will work because webpack looks for all `require` statements. I won't care that it's inside of the `if` statement afaik.

**Comment by gilbox** [comment]
I think [this](https://gist.github.com/defunctzombie/4339901#ignore-a-module) is a solution

**Comment by gilbox** [comment]
```
"browser": {
    "fs": false,
    "path": false
}
```

**Comment by gilbox** [comment]
If that works, it will eliminate errors. However, I think that webpack will compile all language files into the bundle. That's not the intention, right? In the browser you are supposed only include languages that are needed?

**Comment by BenjaminVanRyseghem** [comment]
but if `fs` is not loaded, then `langFile` can't be set, or?

**Comment by gilbox** [comment]
There are several semi-related problems here....

Webpack supports dynamic requires, but it's support is pretty limited. So when it sees this:

```
require(path.join(__dirname, 'languages', langFile))
```

It's first instinct is to try to interpret the intention of the require by analyzing the AST. However, in this case it is too complicated so it throws up it's hands and prints a warning:

```
WARNING in ./~/numbro/numbro.js
Critical dependencies:
970:60-112 the request of a dependency is an expression
 @ ./~/numbro/numbro.js 970:60-112
```

However, it doesn't actually give up at this point. Instead, since it's not sure what files you actually want, it tries to pull in *all* the files in the directory containing `numbro.js`, which leads to the first error I mentioned, because at this point all the files in the directory includes stuff like `.json`, `.md`, `.yml`, files, etc. In theory webpack can actually handle any file type as long as you define a loader for that file type. But, it's unusual to define a loader for an `.md` file unless you are doing something special like generating documentation so I also get errors like this:

```
WARNING in ./~/numbro/CHANGELOG-Numeraljs.md
...
You may need an appropriate loader to handle this file type.
```

(btw, we obviously don't want to include these packages in our bundle so it's good that webpack warns us)

So, my first proposed solution was to do this:

```
require('./languages/' + langFile)
```

... but it's not a great solution because it causes all language files to be included in the bundle. However, it does eliminate all the warnings because webpack understands this expression because it's simple and webpack can convert it into a regex that looks something like `/\.\/languages\/.+/` which it uses to pull in all the files in the `languages/` dir and all those files happen to be `.js` files so webpack doesn't complain.

Anyway, regarding your question:

> but if fs is not loaded, then langFile can't be set, or?

webpack doesn't care if `fs` is loaded or not. It's going to try to analyze all dynamic requires either way.

So, the fix to tell webpack that `fs` is a non-browser related packages is tangential to the fact that webpack wants to analyze a dynamic require which involves the `fs` package.

**Comment by gilbox** [comment]
I can't think of an elegant solution to this problem.

@sokra ?

**Comment by sokra** [comment]
Don't load all languages beforehand. Load them when they are requested. `setLanguage` calls `require("./language/" + lang)` in a try-catch block.

See how moment.js lazy-load locales: https://github.com/moment/moment/blob/develop/src/lib/locale/locales.js#L47

`require("./language/" + lang)` includes all languages in the bundle by default, but you can change this in the webpack configuration: see http://stackoverflow.com/questions/25384360/how-to-prevent-moment-js-from-loading-locales-with-webpack/25426019#25426019

**Comment by BenjaminVanRyseghem** [comment]
That's an elegant solution :)

**Comment by juhaelee** [comment]
any updates on this issue?

**Comment by BenjaminVanRyseghem** [comment]
Not yet.

It's under investigation :smile:

**Comment by Bnaya** [comment]
Same issue here

**Comment by Bnaya** [comment]
I've manged to fix this using the imports lodaer:
```js
module: {
      loaders: [
...
          { test: /numbro\/numbro/,
            loader: "imports?require=>false" },
...
]
}
```

**Comment by jhabdas** [comment]
@Bnaya your loader resolved my issue, so I gave you a :heart:
